consensus/polybft/polybft.go

```diff
@@ -340,124 +307,79 @@ func (p *Polybft) startPbftProcess() {
 	}
 
-	// subscribe to new block events
-	var (
-		newBlockSub   = p.blockchain.SubscribeEvents()
-		syncerBlockCh = make(chan uint64)
-	)
+	newBlockSub := p.blockchain.SubscribeEvents()
+	defer newBlockSub.Close()
+
+	syncerBlockCh := make(chan struct{})
 
-	// Receive a notification every time syncer manages
-	// to insert a valid block.
 	go func() {
 		eventCh := newBlockSub.GetEventCh()
 
 		for {
 			select {
-			case ev := <-eventCh:
-				currentBlockNum := p.blockchain.CurrentHeader().Number
-				if ev.Source == "syncer" {
-					if ev.NewChain[0].Number < currentBlockNum {
-						continue
-					}
-				}
-
-				if p.isSynced() {
-					syncerBlockCh <- currentBlockNum
-				}
-
 			case <-p.closeCh:
 				return
+			case ev := <-eventCh:
+				// The blockchain notification system can eventually deliver
+				// stale block notifications. These should be ignored
+				if ev.Source == "syncer" && ev.NewChain[0].Number > p.blockchain.CurrentHeader().Number {
+					syncerBlockCh <- struct{}{}
+				}
 			}
 		}
 	}()
 
-	defer newBlockSub.Close()
-
-SYNC:
-	if !p.isSynced() {
-		<-syncerBlockCh
-	}
-
-	lastBlock := p.blockchain.CurrentHeader()
-	p.logger.Info("startPbftProcess",
-		"header hash", lastBlock.Hash,
-		"header number", lastBlock.Number)
+	var (
+		sequenceCh   <-chan struct{}
+		stopSequence func()
+	)
 
-	currentValidators, err := p.GetValidators(lastBlock.Number, nil)
-	if err != nil {
-		p.logger.Error("failed to query current validator set", "block number", lastBlock.Number, "error", err)
-	}
+	for {
+		latest := p.blockchain.CurrentHeader().Number
 
-	p.runtime.setIsActiveValidator(currentValidators.ContainsNodeID(p.key.NodeID()))
+		currentValidators, err := p.GetValidators(latest, nil)
+		if err != nil {
+			p.logger.Error("failed to query current validator set", "block number", latest, "error", err)
+		}
 
-	if !p.runtime.isActiveValidator() {
-		// inactive validator is not part of the consensus protocol and it should just perform syncing
-		goto SYNC
-	}
+		isValidator := currentValidators.ContainsNodeID(p.key.NodeID())
+		p.runtime.setIsActiveValidator(isValidator)
 
-	// we have to start the bridge snapshot when we have finished syncing
-	if err := p.runtime.restartEpoch(lastBlock); err != nil {
-		p.logger.Error("failed to restart epoch", "error", err)
+		p.config.TxPool.SetSealing(isValidator) // update tx pool
 
-		goto SYNC
-	}
+		if isValidator {
+			_, err := p.runtime.FSM() // Nemanja: what to do if it is an error
+			if err != nil {
+				p.logger.Error("failed to create fsm", "block number", latest, "error", err)
 
-	for {
-		if err := p.runCycle(); err != nil {
-			if errors.Is(err, errNotAValidator) {
-				p.logger.Info("Node is no longer in validator set")
-			} else {
-				p.logger.Error("an error occurred while running a state machine cycle.", "error", err)
+				continue
 			}
 
-			goto SYNC
+			sequenceCh, stopSequence = p.ibft.runSequence(latest + 1)
+		} else {
+			sequenceCh, stopSequence = nil, nil
```

Comment:
could we remove this `else` by moving `sequenceCh, stopSequence = nil, nil` before the `if`?

Reply:
Fixed [c14c710](https://github.com/0xPolygon/polygon-edge/pull/843/commits/c14c7100feaa8b6d220905a0b02c03632ca046ca) . We have removed else, there is no else in ibft code also. I thought it would be nice to clear sequenceCh if it "points" to previously closed channel but it does not make any difference